{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import os\n",
    "import shutil\n",
    "import pandas as pd\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_tsv_SNVs(filename):\n",
    "    tsv_info = pd.read_csv(filename, sep='\\t', usecols= [0,2,3], header=0)                            \n",
    "    # Extract relevant information from the tsv file\n",
    "    tsv_info['Chromosome'] = tsv_info['DNA Change'].str.split(':').str[0]\n",
    "    tsv_info['Pos'] = tsv_info['DNA Change'].str.split(':').str[1].str.split('>').str[0]\n",
    "    tsv_info['Gene ID'] = tsv_info['Consequences'].str.split().str[1]\n",
    "    tsv_info['Cons_extra'] = tsv_info['Consequences'].str.split().str[2]\n",
    "    tsv_info['Pos'] = tsv_info['Pos'].str.replace('g.', '')\n",
    "    tsv_info['Position'] = tsv_info['Pos'].str[:-1]\n",
    "    tsv_info['Reference'] = tsv_info['Pos'].str[-1]\n",
    "    tsv_info['Alteration'] = tsv_info['DNA Change'].str.split('>').str[1]\n",
    "    tsv_info['fraction'] = tsv_info['# Affected Cases in Cohort'].str.split(',').str[1].str[:-1].astype(float)/100\n",
    "    # create a new dataframe with specific columns from tsv_info\n",
    "    relevant = tsv_info[['Chromosome', 'Position', 'Gene ID', 'Reference', 'Alteration', 'fraction']].copy()\n",
    "    relevant['Position'] = relevant['Position'].astype(int)\n",
    "    relevant['fraction'] = relevant['fraction'].astype(float)\n",
    "    relevant['Chromosome'] = relevant['Chromosome'].str.split('r').str[1]\n",
    "    relevant['Chromosome'] = relevant['Chromosome'].str.replace('X', '30')\n",
    "    #relevant['Chromosome'] = relevant['Chromosome'].str.replace('Y', '31')\n",
    "    relevant['Chromosome'] = relevant['Chromosome'].astype(int)\n",
    "    relevant.sort_values(by=['Chromosome','Position'],inplace=True)\n",
    "    relevant['Chromosome'] = relevant['Chromosome'].astype(str)\n",
    "    relevant['Chromosome'] = relevant['Chromosome'].str.replace('30', 'X')\n",
    "    #relevant['Chromosome'] = relevant['Chromosome'].str.replace('31', 'Y')\n",
    "    relevant = relevant.dropna(axis=0)\n",
    "    relevant.reset_index(drop=True, inplace=True)\n",
    "    # save relevant information to a csv file\n",
    "    relevant.to_csv(\"relevant_info_snvs.csv\",index=False)\n",
    "    return relevant\n",
    "\n",
    "def filter_tsv_Indels(filename):\n",
    "    tsv_info = pd.read_csv(filename, sep='\\t', usecols= [0,2,3], header=0)\n",
    "    # Extract relevant information from the tsv file\n",
    "    tsv_info['Chromosome'] = tsv_info['DNA Change'].str.split(':').str[0]\n",
    "    tsv_info['Gene ID'] = tsv_info['Consequences'].str.split().str[1]\n",
    "    # replace Nans in Gene ID with 'Unknown'\n",
    "    tsv_info['Gene ID'] = tsv_info['Gene ID'].fillna('Unknown')\n",
    "    tsv_info['fraction'] = tsv_info['# Affected Cases in Cohort'].str.split(',').str[1].str[:-1].astype(float)/100\n",
    "    tsv_info['Int'] = tsv_info['DNA Change'].str.split(':').str[1]\n",
    "    tsv_info['Int'] = tsv_info['Int'].str.split('.').str[1]\n",
    "    tsv_info['Pos'] = tsv_info['Int'].str.split('_').str[0]\n",
    "    tsv_info['Position'] = tsv_info['Pos'].str.split('d').str[0]\n",
    "    tsv_info['Remaining1'] = tsv_info['Int'].str.split('del').str[1]\n",
    "    # remove rows where length of Remaining1 is 9\n",
    "    tsv_info = tsv_info[tsv_info['Remaining1'].str.len() != 9]\n",
    "    tsv_info['Remaining1'] = tsv_info['Remaining1'].fillna('-')\n",
    "    tsv_info['Remaining2'] = tsv_info['Int'].str.split('ins').str[1]\n",
    "    tsv_info['Remaining2'] = tsv_info['Remaining2'].fillna('-')\n",
    "    #rename remaining1 and remaining2 to reference and alteration\n",
    "    tsv_info.rename(columns={'Remaining1':'Reference','Remaining2':'Alteration'}, inplace=True)\n",
    "    relevant = tsv_info[['Chromosome', 'Position', 'Gene ID', 'Reference', 'Alteration', 'fraction']].copy()\n",
    "    relevant['Position'] = relevant['Position'].astype(int)\n",
    "    relevant['fraction'] = relevant['fraction'].astype(float)\n",
    "    relevant['Chromosome'] = relevant['Chromosome'].str.split('r').str[1]\n",
    "    relevant['Chromosome'] = relevant['Chromosome'].str.replace('X', '30')\n",
    "    #relevant['Chromosome'] = relevant['Chromosome'].str.replace('Y', '31')\n",
    "    relevant['Chromosome'] = relevant['Chromosome'].astype(int)\n",
    "    relevant.sort_values(by=['Chromosome','Position'],inplace=True)\n",
    "    relevant['Chromosome'] = relevant['Chromosome'].astype(str)\n",
    "    relevant['Chromosome'] = relevant['Chromosome'].str.replace('30', 'X')\n",
    "    #relevant['Chromosome'] = relevant['Chromosome'].str.replace('31', 'Y')\n",
    "    relevant = relevant.dropna(axis=0)\n",
    "    relevant.reset_index(drop=True, inplace=True)\n",
    "    # save relevant information to a csv file\n",
    "    relevant.to_csv(\"relevant_info_indels.csv\",index=False)\n",
    "    return relevant\n",
    "\n",
    "\n",
    "def generate_variants(relevant,num_samples,category):\n",
    "    new_seq = np.empty((relevant.shape[0],num_samples)).astype(str)\n",
    "    for i in tqdm(range(relevant.shape[0])):\n",
    "        rng = np.random.default_rng()\n",
    "        newp = np.array([\"0|0\"]*num_samples)\n",
    "        if relevant['Gene ID'][i] == 'ERBB2':\n",
    "            newp[:] = \"1|1\"\n",
    "        elif relevant['Gene ID'][i] == 'PIK3CA':\n",
    "            if category == 'all':\n",
    "                newp[:] = \"1|1\"\n",
    "            elif category == 'before':\n",
    "                mut = np.arange(0,num_samples,5)\n",
    "                newp[mut] = \"1|1\"\n",
    "            elif category == 'none':\n",
    "                pass\n",
    "                #newp[[j for j in range(num_samples) if j%2==0]] = \"1|1\"\n",
    "        else:\n",
    "            prob = (relevant.iloc[i,-1]*num_samples).astype(np.int32)\n",
    "            mut = rng.choice(num_samples,prob,replace=False)\n",
    "            assert mut.shape[0] == len(set(mut))        \n",
    "            newp[mut] = \"1|1\"\n",
    "        new_seq[i,:] = newp\n",
    "    simulated_seq = pd.DataFrame(new_seq,columns=range(1,num_samples+1))\n",
    "    return simulated_seq\n",
    "\n",
    "def alter_all(relevant,sim_all,num_samples):\n",
    "    sim_before = sim_all.copy()\n",
    "    for i in tqdm(range(relevant.shape[0])):\n",
    "        if relevant['Gene ID'][i] == 'PIK3CA':\n",
    "            newp = np.array([\"0|0\"]*num_samples)\n",
    "            mut = np.arange(0,num_samples,5)\n",
    "            newp[mut] = \"1|1\"\n",
    "            sim_before.iloc[i,:] = newp\n",
    "    return sim_before\n",
    "\n",
    "def create_vcf(subject_index,relevant,sim,startText,num_samples,category):    \n",
    "    vcf_data = \"\"\n",
    "    # add header to vcf_data\n",
    "    ll = len(str(num_samples))\n",
    "    if category == 'none':\n",
    "        subject_ID = str(subject_index+1).zfill(ll)\n",
    "    else:\n",
    "        subject_ID = str(subject_index+(num_samples//2 + 1)).zfill(ll)\n",
    "\n",
    "    if category == 'before':\n",
    "        if subject_index%5 == 0:\n",
    "            vcf_data = startText + \"\\n#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\\t\"+subject_ID+\"_10\\n\"\n",
    "        else:                \n",
    "            vcf_data = startText + \"\\n#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\\t\"+subject_ID+\"_40\\n\"\n",
    "    else:\n",
    "        vcf_data = startText + \"\\n#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\\t\"+subject_ID+\"\\n\"\n",
    "\n",
    "    relevant['Combined'] = relevant['Chromosome'] + \"\\t\" + relevant['Position'].astype(str) + \\\n",
    "    \"\\t\" + relevant['Gene ID'] + \"\\t\" + relevant['Reference'] + \"\\t\" + relevant['Alteration'] + \\\n",
    "    \"\\t.\\t.\\t.\\tGT\\t\" + sim[subject_index+1] #sim[str(subject_index+1)]\n",
    "\n",
    "    vcf_data += \"\\n\".join(relevant['Combined'])\n",
    "    # save vcf file\n",
    "    if category == 'none':\n",
    "        with open(\"all_files/no_pik3ca/subject_\"+subject_ID+\".vcf\", \"w\") as text_file:\n",
    "            text_file.write(vcf_data)\n",
    "    if category == 'all':\n",
    "        with open(\"all_files/both_genes/subject_\"+subject_ID+\".vcf\", \"w\") as text_file:\n",
    "            text_file.write(vcf_data)\n",
    "    elif category == 'before':\n",
    "        if subject_index%5 == 0:\n",
    "            with open(\"all_files/ten_percent/subject_\"+subject_ID+\"_10.vcf\", \"w\") as text_file:\n",
    "                text_file.write(vcf_data)\n",
    "        else:\n",
    "            with open(\"all_files/forty_percent/subject_\"+subject_ID+\"_40.vcf\", \"w\") as text_file:\n",
    "                text_file.write(vcf_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "def tsv_2_vcf(filename1,filename2,startfile,num_samples):\n",
    "    # read relevant_snvs from csv file. If it doesn't exist, create it.\n",
    "    if os.path.isfile(\"relevant_info_snvs.csv\"):\n",
    "        relevant_snvs = pd.read_csv(\"relevant_info_snvs.csv\")\n",
    "    else:\n",
    "        relevant_snvs = filter_tsv_SNVs(filename1)\n",
    "    \n",
    "    # read relevant_indels from csv file. If it doesn't exist, create it.\n",
    "    if os.path.isfile(\"relevant_info_indels.csv\"):\n",
    "        relevant_indels = pd.read_csv(\"relevant_info_indels.csv\")\n",
    "    else:\n",
    "        relevant_indels = filter_tsv_Indels(filename2)\n",
    "\n",
    "    # combine relevant_snvs and relevant_indels into relevant\n",
    "    relevant = pd.concat([relevant_snvs,relevant_indels],ignore_index=True)\n",
    "    relevant.sort_values(by=['Chromosome','Position'],inplace=True)\n",
    "    relevant = relevant_snvs\n",
    "    print(\"Filtering complete\")\n",
    "    r = relevant.copy()\n",
    "\n",
    "    # read header from startfile and save it to a variable\n",
    "    with open(startfile, 'r') as f:\n",
    "        startText = f.read()\n",
    "\n",
    "    os.mkdir(\"all_files\") # create a directory to store all vcf files\n",
    "    os.mkdir(\"all_files/no_pik3ca\") # throws error if folder already exists\n",
    "    os.mkdir(\"all_files/both_genes\") # throws error if folder already exists\n",
    "    os.mkdir(\"all_files/ten_percent\") # throws error if folder already exists\n",
    "    os.mkdir(\"all_files/forty_percent\") # throws error if folder already exists\n",
    "\n",
    "    sim_none = generate_variants(relevant,num_samples//2,'none')\n",
    "    print(\"Creating files with no PIK3CA mutations\")\n",
    "    for i in tqdm(range(num_samples//2)):\n",
    "        create_vcf(i,r,sim_none,startText,num_samples,'none')\n",
    "\n",
    "    sim_all = generate_variants(relevant,num_samples//2,'all')\n",
    "    print(\"Creating files with all PIK3CA mutations\")\n",
    "    for i in tqdm(range(num_samples//2)):\n",
    "            create_vcf(i,r,sim_all,startText,num_samples,'all')\n",
    "\n",
    "    sim_before = alter_all(relevant,sim_all,num_samples//2)\n",
    "    print(\"Creating files where some have PIK3CA mutations and some don't\")\n",
    "    for i in tqdm(range(num_samples//2)):\n",
    "            create_vcf(i,r,sim_before,startText,num_samples,'before')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create vcf files from the given tsv files\n",
    "tsv_2_vcf(filename1='frequent-mutations_GDC_GastricCancer_SNPs.tsv',filename2=\"combined-tsv-files.tsv\",startfile='startText.txt',num_samples=100000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "def reorganize(num_samples,max_files,main_folder):\n",
    "    os.mkdir(main_folder) # throws error if folder already exists\n",
    "    a = int(max_files//14)*14\n",
    "    b = int(a*0.4/1.4)\n",
    "    c = (a-b)//2\n",
    "    d = round(c*2.8)\n",
    "    assert d == a\n",
    "    if num_samples <= c*2:\n",
    "        num_folders = 1\n",
    "    else:\n",
    "        num_folders = num_samples//(c*2)\n",
    "    # Create num_folders number of sub-folders in data_for_babyships folder\n",
    "    for i in range(num_folders):\n",
    "        folder_name = \"set\"+str(i+1)+\"_simulated_gastric_guardant_panel\"\n",
    "        os.mkdir(main_folder+\"/\"+folder_name)\n",
    "\n",
    "    # Moving all files to the correct folder\n",
    "    for i in range(num_folders):\n",
    "        folder_name = \"set\"+str(i+1)+\"_simulated_gastric_guardant_panel\"\n",
    "\n",
    "        # Copy every c files from no_pik3ca folder to the new folders\n",
    "        for file in os.listdir(\"all_files/no_pik3ca\"):\n",
    "            if file != \".DS_Store\":\n",
    "                if i*c < int(file.split('.')[0].split('_')[1]) <= (i+1)*c:\n",
    "                    shutil.copy(\"all_files/no_pik3ca/\"+file,main_folder+\"/\"+folder_name+\"/\"+file)\n",
    "        \n",
    "        # Copy every c files from both_genes folder to the new folders\n",
    "        for file in os.listdir(\"all_files/both_genes\"):\n",
    "            if file != \".DS_Store\":\n",
    "                if i*c < int(file.split('.')[0].split('_')[1])-num_samples//2 <= (i+1)*c:\n",
    "                    shutil.copy(\"all_files/both_genes/\"+file,main_folder+\"/\"+folder_name+\"/\"+file)\n",
    "\n",
    "        # Repeat the above steps for the 40 percent folder    \n",
    "        for file in os.listdir(\"all_files/forty_percent\"):\n",
    "            if file != \".DS_Store\":\n",
    "                if i*c < int(file.split('_')[1])-num_samples//2 <= (i+1)*c:\n",
    "                    shutil.copy(\"all_files/forty_percent/\"+file,main_folder+\"/\"+folder_name+\"/\"+file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "reorganize(num_samples=100000, max_files=23000, main_folder = \"data_for_babyships_5\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.12 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "73d4432868320a00a661825a130a6a730fb841689da4e88a956d3ce9f1015fce"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
